compiler/rustc_typeck/src/check/fn_ctxt/checks.rs

```diff
@@ -236,137 +142,640 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                    }
+                    // If we eliminate the last column, any left-over arguments are extra
+                    if mat[i].len() == 0 {
+                        return Some(Issue::Extra(i));
+                    }
 
-                let is_closure = matches!(arg.kind, ExprKind::Closure(..));
+                    // Make sure we don't pass the bounds of our matrix
+                    let is_arg = i < ai.len();
+                    let is_input = i < ii.len();
+                    if is_arg && is_input && mat[i][i] {
+                        // This is a satisfied input, so move along
+                        continue;
+                    }
 
-                if is_closure != check_closures {
-                    continue;
+                    let mut useless = true;
+                    let mut unsatisfiable = true;
+                    if is_arg {
+                        for j in 0..ii.len() {
+                            // If we find at least one input this argument could satisfy
+                            // this argument isn't completely useless
+                            if mat[i][j] {
+                                useless = false;
+                                break;
+                            }
+                        }
+                    }
+                    if is_input {
+                        for j in 0..ai.len() {
+                            // If we find at least one argument that could satisfy this input
+                            // this argument isn't unsatisfiable
+                            if mat[j][i] {
+                                unsatisfiable = false;
+                                break;
+                            }
+                        }
+                    }
+
+                    match (is_arg, is_input, useless, unsatisfiable) {
+                        // If an input is unsatisfied, and the argument in it's position is useless
+                        // then the most likely explanation is that we just got the types wrong
+                        (true, true, true, true) => return Some(Issue::Invalid(i)),
+                        // Otherwise, if an input is useless, then indicate that this is an extra argument
+                        (true, _, true, _) => return Some(Issue::Extra(i)),
+                        // Otherwise, if an argument is unsatisfiable, indicate that it's missing
+                        (_, true, _, true) => return Some(Issue::Missing(i)),
+                        (true, true, _, _) => {
+                            // The argument isn't useless, and the input isn't unsatisfied,
+                            // so look for a parameter we might swap it with
+                            // We look for swaps explicitly, instead of just falling back on permutations
+                            // so that cases like (A,B,C,D) given (B,A,D,C) show up as two swaps,
+                            // instead of a large permutation of 4 elements.
```

Comment:
We already limit it to mentioning the arguments that need to be changed, but the problem is the permutation error message is going to be denser, because you kind of have to squint to understand the original order and the proposed order.  If we can decompose the permutation into a set of independent swaps, it becomes much easier to understand (at least for me).

Permutation is really meant to capture the case where you've shuffled a large number of arguments (and, arguably, we might want to just end up falling back to reporting that the whole call is invalid)